src/module/iotjs_module_uart.c

```diff
@@ -0,0 +1,238 @@
+  iotjs_jobjectwrap_destroy(&_this->jobjectwrap);
+  IOTJS_RELEASE(uart);
+}
+
+
+const iotjs_jval_t* iotjs_uart_get_juart() {
+  return iotjs_module_get(MODULE_UART);
+}
+
+
+iotjs_uart_t* iotjs_uart_get_instance() {
+  const iotjs_jval_t* juart = iotjs_uart_get_juart();
+  iotjs_jobjectwrap_t* jobjectwrap = iotjs_jobjectwrap_from_jobject(juart);
+  return (iotjs_uart_t*)jobjectwrap;
+}
+
+void iotjs_uart_onread(iotjs_jval_t* jthis, char* buf) {
+  iotjs_jval_t jemit = iotjs_jval_get_property(jthis, "emit");
+  IOTJS_ASSERT(iotjs_jval_is_function(&jemit));
+
+  iotjs_jargs_t jargs = iotjs_jargs_create(2);
+  iotjs_jval_t str = iotjs_jval_create_string_raw("data");
+  iotjs_jval_t data = iotjs_jval_create_string_raw(buf);
+  iotjs_jargs_append_jval(&jargs, &str);
+  iotjs_jargs_append_jval(&jargs, &data);
+  iotjs_jhelper_call_ok(&jemit, jthis, &jargs);
+
+  iotjs_jval_destroy(&str);
+  iotjs_jval_destroy(&data);
+  iotjs_jargs_destroy(&jargs);
+}
+
+
+void AfterUARTWork(uv_work_t* work_req, int status) {
+  iotjs_uart_t* uart = iotjs_uart_get_instance();
+
+  iotjs_uart_reqwrap_t* req_wrap = iotjs_uart_reqwrap_from_request(work_req);
+  iotjs_uart_reqdata_t* req_data = iotjs_uart_reqwrap_data(req_wrap);
+
+  iotjs_jargs_t jargs = iotjs_jargs_create(2);
+
+  // TODO
+  if (status) {
+    iotjs_jval_t error = iotjs_jval_create_error("System error");
+    iotjs_jargs_append_jval(&jargs, &error);
+    iotjs_jval_destroy(&error);
+  } else {
+    switch (req_data->op) {
+      case kUartOpOpen: {
+        if (req_data->error == kUartErrOpen) {
+          iotjs_jval_t error =
+              iotjs_jval_create_error("Failed to open UART device");
+          iotjs_jargs_append_jval(&jargs, &error);
+          iotjs_jval_destroy(&error);
+        } else {
+          iotjs_jargs_append_null(&jargs);
+        }
+        break;
+      }
+      case kUartOpWrite: {
+        if (req_data->error == kUartErrWrite) {
+          iotjs_jval_t error =
+              iotjs_jval_create_error("Cannot write to device");
+          iotjs_jargs_append_jval(&jargs, &error);
+          iotjs_jval_destroy(&error);
+        } else {
+          iotjs_jargs_append_null(&jargs);
+        }
+        break;
+      }
+      default: {
+        IOTJS_ASSERT(!"Unreachable");
+        break;
+      }
+    }
+  }
+
+  const iotjs_jval_t* jcallback = iotjs_uart_reqwrap_jcallback(req_wrap);
+  const iotjs_jval_t* juart = iotjs_uart_get_juart();
+  iotjs_make_callback(jcallback, juart, &jargs);
+
+  iotjs_jargs_destroy(&jargs);
+  iotjs_uart_reqwrap_dispatched(req_wrap);
+}
+
+
+#define UART_ASYNC(op)                                                 \
+  do {                                                                 \
+    uv_loop_t* loop = iotjs_environment_loop(iotjs_environment_get()); \
+    uv_work_t* req = iotjs_uart_reqwrap_req(req_wrap);                 \
+    uv_queue_work(loop, req, op##WorkerUart, AfterUARTWork);           \
```

Comment:
It's because I found out that without adding Uart after Worker, it makes a conflict with Worker functions in i2c module